test/zdtm/static/pidfd_self.c

```diff
@@ -0,0 +1,153 @@
+#include <stdio.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/syscall.h>
+#include <signal.h>
+#include <unistd.h>
+
+#include "zdtmtst.h"
+
+const char *test_doc = "Check pidfd /proc/self/fdinfo/<pidfd> entry remains consistent after checkpoint/restore\n";
+const char *test_author = "Bhavik Sachdev <b.sachdev1904@gmail.com>";
+
+struct pidfd_status {
+	long pos;
+	unsigned int flags;
+	int mnt_id;
+	int ino;
+	pid_t pid;
+};
+
+static int pidfd_open(pid_t pid, unsigned int flags)
+{
+	return syscall(__NR_pidfd_open, pid, flags);
+}
+
+static int pidfd_send_signal(int pidfd, int sig, siginfo_t* info, unsigned int flags)
+{
+	return syscall(__NR_pidfd_send_signal, pidfd, sig, info, flags);
+}
+
+static void show_pidfd(char *prefix, struct pidfd_status *s)
+{
+	test_msg("%s\n\tpos: %ld\n\tflags: 0%o\n\tmnt_id: %d\n\tino: %d\n\tpid: %d\n",
+	prefix, s->pos, s->flags, s->mnt_id, s->ino, s->pid);
+}
+
+static int parse_self_fdinfo(int pidfd, struct pidfd_status *s)
+{
+	char buf[256];
+	int ret = -1;
+	FILE *f;
+
+	sprintf(buf, "/proc/self/fdinfo/%d", pidfd);
+	f = fopen(buf, "r");
+	if (!f) {
+		perror("Can't open /proc/self/fdinfo/ to parse");
+		return -1;
+	}
+
+	memset(s, 0, sizeof(*s));
+
+	/*
+	* pos:	  current file offset
+	* flags:  file access mode (octal) 02000002 => [O_RDWR | O_CLOEXEC]
+	* mnt_id: mount id associated with the file descriptor
+	* ino:    inode
+	* NSpid:  pid in the namespace the process is in
+	* pid:    it's pid
+	*/
+	while (fgets(buf, sizeof(buf), f)) {
+		if (sscanf(buf, "pos: %li", &s->pos) != 1)
+			goto parse_err;
+
+		if (!fgets(buf, sizeof(buf), f))
+			goto parse_err;
+
+		if (sscanf(buf, "flags: 0%o", &s->flags) != 1)
+			goto parse_err;
+
+		if (!fgets(buf, sizeof(buf), f))
+			goto parse_err;
+
+		if (sscanf(buf, "mnt_id: %d", &s->mnt_id) != 1)
+			goto parse_err;
+
+		if (!fgets(buf, sizeof(buf), f))
+			goto parse_err;
+
+		if (sscanf(buf, "ino: %d", &s->ino) != 1)
+			goto parse_err;
+
+		if (!fgets(buf, sizeof(buf), f))
+			goto parse_err;
+
+		if (sscanf(buf, "Pid: %d", &s->pid) != 1)
+			goto parse_err;
+		ret = 0;
+		break;
+	}
+
+	if (ret)
+		goto parse_err;
+err:
+	fclose(f);
+	return ret;
+
+parse_err:
+	perror("Format error");
+	goto err;
+}
+
+static int check_pidfd(int fd, struct pidfd_status *old)
+{
+	struct pidfd_status new;
+
+	if (parse_self_fdinfo(fd, &new))
+		return -1;
+
+	show_pidfd("restored", &new);
+
+	if (old->pos != new.pos || old->flags != new.flags ||
+		old->mnt_id != new.mnt_id || old->ino != new.ino
```

Comment:
we don't need to compare `mnt_id` and `pos`

Reply:
Done!